### PR TITLE
ARROW-4699: [C++] remove json chunker's requirement of null terminated buffers

### PR DESCRIPTION
**Pull request by bkietz**
Re-enable json chunker test. [This was failing valgrind](https://travis-ci.org/apache/arrow/jobs/494039471#L2140) but that is not currently happening locally

### CODE REVIEW

**Comment by xhochy** [comment]
@bkietz Tests are now failing in Travis.

**Comment by bkietz** [comment]
I don't know what caused the test failure. I have rebased and the remaining Travis failures seem unrelated:
[R](https://travis-ci.org/apache/arrow/jobs/502215724#L973), [GCC 4.8](https://travis-ci.org/apache/arrow/jobs/502215705#L496-L501)

**Comment by kszucs** [comment]
Restarted them.

File: cpp/src/arrow/json/chunker.cc [hunk 1]
```diff
@@ -167,7 +167,9 @@ class ParsingChunker : public Chunker {
     }
     std::size_t total_length = 0;
     for (auto consumed = block;; consumed = block.substr(total_length)) {
-      auto length = ConsumeWholeObject(rapidjson::StringStream(consumed.data()));
+      rapidjson::MemoryStream ms(consumed.data(), consumed.size());
+      auto length = ConsumeWholeObject(
+          rapidjson::EncodedInputStream<rapidjson::UTF8<>, rapidjson::MemoryStream>(ms));
```
Comment (fsaintjacques) [inline cpp/src/arrow/json/chunker.cc]: I'd use `using InputStream = rapid...` just for making this fit on one line.

**Review by fsaintjacques — CHANGES_REQUESTED** [review]
Seems like the travis failure is related.

File: cpp/src/arrow/json/chunker-test.cc [hunk 2]
```diff
@@ -49,9 +49,9 @@ std::string join(Lines&& lines, std::string delimiter) {
   return joined;
 }
 
-std::string PrettyPrint(std::string one_line) {
```
Comment (bkietz) [inline cpp/src/arrow/json/chunker-test.cc]: While debugging this I found that GCC 4.8 was not making a deep copy of this argument, which resulted in the corruption of test data for tests following `ChunkerTest.PrettyPrinted`

Comment (bkietz) [reply cpp/src/arrow/json/chunker-test.cc]: I guess the implementation was copy-on-write?

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**